Bug 1064144 - Improve focus behaviour when opening a draft

autra opened the pull request.

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9341322cde15ea07057951581865518784b70c09

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=dc81b208460c25abf55f6d86423c1df3cf101ddb

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7a5dfbea1c0c35283156f3181006f74231fedcb4

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=655155f97e341367982d6e72526a41266e52c862

autra committed:
Bug 1064144 - Improve focus behaviour when opening a draft

Add a 'focus' arg to afterEnter:
- If its value is 'composer', focus the composer
- If its value is 'recipients', focus the recipients.
- If any other value, focus the recipients if there is none yet, and focus the
composer if there are some.

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=dd85b89e0435d7321853d1f6814f6f1ae3a66206

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b05989d7f794185aec0113470bdf3042b5421a70

autra commented:
So there are 2 commits, the second one only fixing some trailing whitespaces. I included it because my vim keeps fixing them, so I thought I'd better commit it anyway. I don't know what is the policy for this kind of fixes in gaia...

autra commented on apps/sms/views/conversation/js/conversation.js:
```diff
@@ -769,10 +779,11 @@ var ConversationView = {
       // Otherwise find them from draft recipient numbers
       this.draft.recipients.forEach(function(number) {
-        Contacts.findByAddress(number).then(function(contacts) {
+        promises.push(Contacts.findByAddress(number).then(function(contacts) {
```
this `then` callback was executed after `draftPromise.then` line 648. In other word, now that we check the recipients to decide where to focus, we need to wait for their initialization

julienw replied:
yep :) but you can use "this.draft.recipients.map" (instead of forEach) and you don't need to use the array and push in it.

julienw replied:
and while you're here, maybe convert the functions to arrow functions, and remove the "bind" calls

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=3e5e340b505c3705e423b609a5f29df34be091b2

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=464db92c6678faff65f070017069dd0a94f72702

autra commented on apps/sms/test/marionette/drafts_test.js:
```diff
@@ -283,6 +289,34 @@ marionette('Messages Drafts', function() {
       });
     });
 
+    suite('Thread-less recipient-less drafts', function() {
```
I put this test in its own suite. Didn't find any easy way to include that in `Thread-less drafts` test suite... Do you have a better idea?

julienw replied:
I think it's ok here.

azasypkin replied:
__nit:__ and please ```Conversation-less drafts``` :) I noticed that new contributors are confused by the fact that we use both ```conversation``` and ```thread``` across our codebase.

autra replied:
Oh. So there's a difference between conversation and thread? :-)

azasypkin replied:
I think they are the same, also in inbox-view terms conversation-less draft is "conversation" as well, @julienw please correct me if I'm wrong :)

julienw replied:
we changed our terminology from "thread" to "conversation" as this is more understandable and less overloaded...

In Inbox "conversation-less draft" is an item of the list, don't know if we should call this a "conversation" but at least it has the same info :) maybe in Inbox we should use the more general term "inbox item" :)

azasypkin replied:
> In Inbox "conversation-less draft" is an item of the list, don't know if we should call this a "conversation" but at least it has the same info :) maybe in Inbox we should use the more general term "inbox item" :)

Heh, note that we're going to return both conversations and conversation-less drafts as "conversation/conversation-info" via ```ConversationService.getAllConversations``` :) [1]

[1] https://github.com/mozilla-b2g/gaia/blob/master/apps/sms/services/README.md

julienw replied:
yeah I know; maybe we can find better semantics

julienw commented on apps/sms/views/conversation/js/conversation.js:
```diff
@@ -642,10 +642,20 @@ var ConversationView = {
+          this.recipients.focus();
+          break;
+        default:
+          if (this.hasValidRecipients()) {
```
right now we don't save invalid recipients, but we will eventually. Wondering what we should do if we have only invalid recipients... I'd say... we'll see when it will happen :p

autra replied:
If I understand correctly, `this.hasValidRecipients` only checks that the input value is not empty and finite right? So actually, it actually checks whether or not we have something.In case of only invalid recipients, maybe we'd want to focus the recipients input instead of the composer... Yep, we'll see what happens :-p

julienw commented on apps/sms/test/marionette/lib/views/new-message/view.js:
```diff
@@ -167,6 +167,12 @@ NewMessageView.prototype = {
+  isRecipientsInputFocused: function() {
+    return this.accessors.recipientsInput.scriptWith(function(el) {
+      return document.activeElement === el;
+    });
```
wondering if we should have a more generic "isElementFocused" that we could reuse. I don't know where we can have such a toolset though... @azasypkin what do you think ?

azasypkin replied:
Yeah, would be great to have shared tools for that and we don't have them now :/ What about ```view_tools (or better name).js``` in [1]?

[1] https://github.com/mozilla-b2g/gaia/tree/master/apps/sms/test/marionette/lib/views/shared

julienw commented on apps/sms/views/conversation/test/unit/conversation_test.js:
```diff
@@ -6786,6 +6783,8 @@ suite('conversation.js >', function() {
         Navigation.isCurrentPanel.withArgs('composer').returns(true);
         this.sinon.stub(ConversationView.recipients, 'focus');
+        this.sinon.stub(Compose, 'focus');
+        this.sinon.stub(ConversationView, 'hasValidRecipients');
```
I don't think it's a valid way to stub internal methods. Can you write tests that actually add drafts with/without recipients instead ?

autra replied:
Could you elaborate why it is not valid? What do you mean by that?

Actually here, as we are unit-testing, isn't it better to treat results of other functions as part of the external state, and therefore to mock them, just like parameters? That's usually how I write my unit test (but are we really unit-testing here?) . WDYT?

julienw replied:
In the term "unit testing", we're testing a "unit of code"; the unit of code is the whole object ConversationView.
So ideally we should not mock other methods of this object while testing a method of this object. We don't need to test how the unit works internally, this is not what's important. What's important is how the unit works with inputs/outputs.

Also this makes it easier to refactor the unit.

if we feel like we need to, it usually means the object holds several responsiblities and we should split it.

In this case, I don't think it's the case though. It's just as easy to add something (or not) to ConversationView.recipients.numbers (I think ConversationView.recipients is a MockRecipients, but please check this).

Hope this is clearer !

autra replied:
Completely clear and making sense, thanks!

julienw commented on apps/sms/views/shared/js/activity_handler.js:
```diff
@@ -211,22 +211,22 @@ var ActivityHandler = {
         return Navigation.toPanel('thread', {
           id: message.threadId,
-          focusComposer: focusComposer
+          focus: focus
```
nit: use just the shortcut "focus"

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=33aaa253a4c93a0dad77b048469c42a8396c891f

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e77d8fb29484f3aa82be2d7e4812e9c44383b43f

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=1d1e05128db19b15e1cc56b082c119eafe33ffe5

autra commented on apps/sms/test/marionette/drafts_test.js:
```diff
@@ -200,7 +206,7 @@ marionette('Messages Drafts', function() {
     });
 
-    suite('Thread-less drafts', function() {
+    suite('Conversation-less drafts', function() {
```
So renaming also the other `Thread` of this test :-)

autra commented on apps/sms/test/marionette/lib/views/conversation/view.js:
```diff
@@ -6,6 +6,7 @@ var ConversationAccessor = require('./accessors');
 var MessageAccessor = require('./message_accessors');
 var ComposerAccessor = require('../shared/composer_accessors');
 var MenuAccessor = require('../shared/menu_accessors');
+var Tools = require('../shared/tools.js');
```
Here I'm not really sure I've done things the way you expected. This certainly allows to factorize the `scriptWith...` stuff, but I feel I should've gone farther. 

On the other hand, if I remove or move the `isMessageInputFocused` and `isRecipientInputFocused`, I would need to declare `messageAccessor` and `composerAccessor` in `Tools` and I didn't feel that worthy. WDYT?

autra replied:
Actually I realised I forgot some isElementFocused in individual tests. Will update.

autra commented on apps/sms/test/marionette/lib/views/shared/tools.js:
```diff
@@ -0,0 +1,16 @@
+/* global module */
+
+function Tools() {
+}
```
I created a constructor here like in other files, but is it ok to just export a single object containing my method?

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d1327f79e61b573450773fb421f95f06934264cc

julienw commented:
> I created a constructor here like in other files, but is it ok to just export a single object containing my method?

I actually prefer it like you did in the last commit :)

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a9947831a46976d017a7f7ecf2f9c656fcfad28c

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e7e58954b53dd9312034ff129e6a27abe7af47e7
